CarAccidentErrorHandling/CarAccidentErrorHandling.ts

```diff
@@ -0,0 +1,118 @@
+class SlipperyRoadAccidentException extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+class CollisionAccidentException extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+class HitAndRunAccidentException extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+class EnvironmentalAccidentException extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+function driveSafely() {
+  const isSlipperyRoad = true;
+  const isCollision = false;
+  const isEnvironmental = false;
+
+  if (isSlipperyRoad) {
+    throw new SlipperyRoadAccidentException(
+      "Slippery road due to rain. Drive with caution."
+    );
+  } else if (isCollision) {
+    throw new CollisionAccidentException(
+      "Collision with another vehicle occurred."
+    );
+  } else if (isEnvironmental) {
+    throw new EnvironmentalAccidentException(
+      "Environmental conditions such as thunderstorm. Drive with caution."
+    );
+  } else {
+    throw new HitAndRunAccidentException("Hit and run incident happened.");
+  }
+}
+
+function handleSlipperyRoadAccident() {
+  callEmergencyServices();
+  callTowTruck();
+}
+
+function handleCollisionAccident() {
+  callEmergencyServices();
+  assessInjuries();
+}
+
+function handleHitAndRunAccident() {
+  callEmergencyServices();
+  inspectDamage();
+  notifyAuthorities();
+}
+
+function handleEnvironmentalAccident() {
+  callEmergencyServices();
+  seekShelter();
+}
+
+function callEmergencyServices() {
+  console.log("Emergency services are being contacted.");
+}
+
+function callTowTruck() {
+  console.log("Tow truck is on the way to clear the vehicle.");
+}
+
+function assessInjuries() {
+  console.log(
+    "Injuries are being assessed, and medical assistance is provided if needed."
+  );
+}
+
+function inspectDamage() {
+  console.log("Damage to the vehicle is being inspected.");
+}
+
+function notifyAuthorities() {
+  console.log("Authorities are being notified about the hit and run incident.");
+}
+
+function seekShelter() {
+  console.log("Seeking shelter due to environmental conditions.");
+}
+
+function cleanUpAfterAccident() {
+  console.log("Cleanup procedures initiated.");
+}
+
+try {
+  driveSafely();
+} catch (e) {
+  if (e instanceof SlipperyRoadAccidentException) {
+    handleSlipperyRoadAccident();
+  } else if (e instanceof CollisionAccidentException) {
+    handleCollisionAccident();
+  } else if (e instanceof EnvironmentalAccidentException) {
+    handleEnvironmentalAccident();
+  } else if (e instanceof HitAndRunAccidentException) {
+    handleHitAndRunAccident();
+  } else {
+    console.error("An error occurred:", e.message);
+  }
+} finally {
+  cleanUpAfterAccident();
```

Comment:
what type of cleanup are we referring to here, could be more specific.

Reply:
Thanks for highlighting. Added the function in the catch block itself.

Reply:
Changed the name to something more meaningful =>initiateRoadCleanup